<a href= "https://colab.research.google.com/github/Nkhanal2048/Artificial-Intelligence/blob/main/AI_assignment%234.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


In [ ]:

# Imports and pip installations (if needed)

import pandas as pd
import numpy as np


# Part 1: Load the dataset

In [4]:
# Load the dataset (load remotely, not locally)

from sklearn.datasets import load_iris
iris = load_iris()
iris_df = pd.DataFrame(iris['data'], columns = iris.feature_names) # transforming load_iris dictionary into a dataset
iris_df["species"] = iris.target


In [5]:
# Output the first 15 rows of the data
iris_df.head(15)


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [6]:
# Display a summary of the table information (number of datapoints, etc.
iris_df.info() # for technical summary
iris_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
sepal length (cm)    150 non-null float64
sepal width (cm)     150 non-null float64
petal length (cm)    150 non-null float64
petal width (cm)     150 non-null float64
species              150 non-null int64
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)     species  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

In [7]:
# number of datapoints
iris_df.size 

750

About the dataset

Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

Answer: This is a multivariable data set exploring linear discrimmination analysis. It deals with sepal length, sepal width, petal length, petal width and 3 species of Iris. (Paraphrased the Wikipedia article).

Our features are the aforementioned "sepal length (cm)", "sepal width (cm)", "petal length (cm)", and "petal width (cm)"; and our label is "species".

Mapping: we are looking at each features and their values to determine what species they best match up with.

For example, in the first row we have: Sepal length: 5.1, Sepal width: 3.5, Petal length: 1.4, Petal width: 0.2 which determines that our species is "0" or "setosa".



# Part 2: Split the dataset into train and test

In [36]:
# Take the dataset and split it into our features (X) and label (y)

features = iris_df[["sepal length (cm)", "sepal width (cm)", "petal length (cm)", "petal width (cm)"]] # X (2D array)
label = iris_df["species"] # y


In [10]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

from sklearn.model_selection import train_test_split

# splits into 4 arrays
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9)

In [8]:
X_train, X_test, y_train, y_test # for display 


(     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
 120                6.9               3.2                5.7               2.3
 100                6.3               3.3                6.0               2.5
 42                 4.4               3.2                1.3               0.2
 41                 4.5               2.3                1.3               0.3
 48                 5.3               3.7                1.5               0.2
 ..                 ...               ...                ...               ...
 63                 6.1               2.9                4.7               1.4
 95                 5.7               3.0                4.2               1.2
 69                 5.6               2.5                3.9               1.1
 85                 6.0               3.4                4.5               1.6
 5                  5.4               3.9                1.7               0.4
 
 [135 rows x 4 columns],
      sepal length (cm)  

# Part 3: Logistic Regression

In [11]:
# i. Use sklearn to train a LogisticRegression model on the training set

from sklearn.linear_model import LogisticRegression

LogReg = LogisticRegression(max_iter = 500).fit(X_train, y_train) # for fitting; use train. 




/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


In [12]:
# ii. For a sample datapoint, predict the probabilities for each possible class

pd.DataFrame(LogReg.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"]) # for the whole X_test



Setosa  Versicolor  Virginica
0  0.001006    0.002381   0.996613

In [13]:
# But now I'll test sample values from Wikipedia's chart to see if they correlerate with the correct species.
LogReg.predict(np.array([[4.9, 3.1, 1.5, 0.1]])) # 0 meaning Setosa **

array([0])

In [14]:
LogReg.predict(np.array([[4.9, 2.4, 3.3, 1.0]])) # 1 meaning Versicolor

array([1])

In [15]:
LogReg.predict(np.array([[6.4, 2.8, 5.6, 2.1]])) # 2 meaning Virginica


array([2])

In [16]:
# iii. Report on the score for Logistic regression model, what does the score measure?

LogReg.score(X_test, y_test) # for scoring; use test.


0.8666666666666667

The 0, 1, and 2 from the predictions correspond to Setosa, Versicolor, and Virginica respectively.

The 0.8666666666666667 score means there's an 86% chance of this classification being correct.

In [17]:
# iv. Extract the coefficents and intercepts for the boundary line(s)

LogReg.coef_
#LogReg.intercept_

array([[ 0.39976589,  1.42824402, -2.20344357, -0.95200226],
       [ 0.55759422, -1.72872263,  0.48452054, -1.3813891 ],
       [-1.74390749, -1.28837152,  2.40284043,  2.43947634]])

In [18]:
LogReg.intercept_

array([ 0.26457057,  0.93751908, -1.1374495 ])

# Part 4: Support Vector Machine

In [42]:
# i. Use sklearn to train a Support Vector Classifier on the training set
from sklearn import svm

#model = svm.SVC(kernel = "linear")
#model.fit(X_train, y_train)
#svm.SVC(LogReg, X_train, y_train)

svm_model = svm.SVC(probability = True).fit(features, label)


/opt/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


In [20]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(svm_model.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])  # for the entire X_test


Setosa  Versicolor  Virginica
0  0.240533    0.286253   0.473215

In [21]:
# Like before, I'll test sample values from Wikipedia's chart to see if they correlerate with the correct species.
svm_model.predict(np.array([[4.9, 3.1, 1.5, 0.1]])) # 0 meaning Setosa

array([0])

In [22]:
svm_model.predict(np.array([[4.9, 2.4, 3.3, 1.0]])) # 1 meaning Versicolor

array([1])

In [23]:
svm_model.predict(np.array([[6.4, 2.8, 5.6, 2.1]])) # 2 meaning Virginica


array([2])

In [24]:
# iii. Report on the score for the SVM, what does the score measure?
svm_model.score(X_test, y_test)

1.0

# Part 5: Neural Network

In [25]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
from sklearn.neural_network import MLPClassifier

mlp_model_lbfgs = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(mlp_model_lbfgs.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])



Setosa  Versicolor     Virginica
0     0.0         1.0  4.245555e-34

In [26]:
# Like before, I'll test sample values from Wikipedia's chart to see if they correlerate with the correct species.
mlp_model_lbfgs.predict(np.array([[4.9, 3.1, 1.5, 0.1]])) # array 0 is Setosa


array([0])

In [27]:
mlp_model_lbfgs.predict(np.array([[4.9, 2.4, 3.3, 1.0]])) # array 1 is Versicolor


array([1])

In [28]:
mlp_model_lbfgs.predict(np.array([[6.4, 2.8, 5.6, 2.1]])) # array 2 is Virginica


array([2])

In [29]:
# iii. Report on the score for the Neural Network, what does the score measure?
mlp_model_lbfgs.score(X_test, y_test)


1.0

The 0, 1, and 2 from the predictions correspond to Setosa, Versicolor, and Virginica respectively.

The 1 score means there's an 100% chance of this classification being correct.



In [30]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)
mlp_model_sgd = MLPClassifier(max_iter = 1500, solver='sgd', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)

In [31]:
pd.DataFrame(mlp_model_sgd.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])


Setosa  Versicolor  Virginica
0  0.000004    0.000076   0.999921

In [32]:
mlp_model_sgd.score(X_test, y_test) 

1.0

I got 1 in all three cases.

In [33]:
mlp_model_adam = MLPClassifier(max_iter = 1550, solver='adam', alpha=1e-5, hidden_layer_sizes=(15,), random_state=1).fit(features, label)


In [34]:
pd.DataFrame(mlp_model_adam.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])


Setosa  Versicolor  Virginica
0  0.000001    0.000001   0.999998

In [35]:
mlp_model_adam.score(X_test, y_test)


1.0

# Part 6: K-Nearest Neighbors

In [38]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

from sklearn.neighbors import KNeighborsClassifier

KNN_model = KNeighborsClassifier(n_neighbors=3).fit(X_train, y_train) # for fitting; use train. *limit on max inters




In [39]:
# ii. For a sample datapoint, predict the probabilities for each possible class
pd.DataFrame(KNN_model.predict_proba([[1, 2, 3, 4]]), columns = ["Setosa", "Versicolor", "Virginica"])


Setosa  Versicolor  Virginica
0     0.0    0.666667   0.333333

In [40]:
# iii. Report on the score for kNN, what does the score measure?
KNN_model.score(X_test, y_test)

1.0

# Part 7: Conclusions and takeaways

<div>
In your own words describe the results of the notebook. 
Which model(s) performed the best on the dataset? Why do you think that is?
Did anything surprise you about the exercise?

Answer:

1.Logistic: 
    probabilties 0.315912 0.098862 0.585226 score: 0.8666666666666667

2. SVM: 
    probabilities: 0.251848 0.237385 0.510768 score: 0.9333333333333333
 <br>

3. Neural Network:
    lbfgs probabilties: 0.0 1.0 9.018592e-15 score: 1

    sgd probabilties: 0.000004 0.000076 0.999921 score: 1

    adam probabilities: 0.000001 0.000001 0.999998 score: 1

    KNN probabilities: 0.0 1.0 0.0 score: 0.9333333333333333

It seems like Neural Networks performed the best on the dataset.

But I can see different scores upon each run from all models getting the same score (either all 1s, 86%, and 93% etc).

I assume this is due to reshuffling as I rerun my code several times and used max_iter where appropriate, sometimes I got all 1s which is possibility but doesn't seem right? 
I would also get 86% and now I got 93%.
</div>
